### ISSUE TITLE
Is verification inefficient when there are many trust anchors?

### ISSUE THREAD

**Issue by briansmith** [briansmith/webpki#38]
In https://github.com/ctz/rustls/pull/47#issue-204641103, @avadacatavra published some numbers that showed that doing a bunch of signature verifications in parallel (each with only one trust anchor?) was significantly faster than doing a single verification with a large number of trust anchors (from https://github.com/ctz/webpki-roots?). The point of this issue is to investigate why that is the case, and identify potential solutions.

https://github.com/ctz/rustls/pull/47#issuecomment-276959262 lists 13 sites that are being used by the Servo project to benchmark performance. I think it should be straightforward to extract the certificate chains they are sending and build a benchmark using those certificate chains + webpki-roots, but which is independent of a TLS stack, to get a rough idea of what the issues are, regardless of whether we think that corpus is the right one for us to optimize for.

Using the benchmark, we should be able to profile performance and find out the issue.

Speculation: webpki wasn't really written with the expectation that a large number of trust anchors would be used. Instead, it was originally designed for embedded applications which would have only a few trust anchors. webpki contains an optimization for trying to build a short chain, which is simply looping over the trust anchors before looping over the given intermediates. This is a linear search, which is reasonable if the number of trust anchors is small, but maybe not if the number of trust anchors isn't small. If the profile shows that the `match loop_while_non_fatal_error(trust_anchors,` expression is taking a significant amount of time (note: it may be easier to see this if each `loop_while_non_fatal_error(...` call is factored out into its own function), then we can optimize that. For example, we could use https://github.com/sfackler/rust-phf to replace the linear search with a hash table lookup, if webpki-roots would be willing to implement a new API based on that.

@avadacatavra Is my description of your measurements (using webpki-roots as the root store, and benchmarking many threads using 1 trust anchor vs. 1 thread using all of the trust anchors in webpki-roots) accurate?

/cc @ctz

**Comment by ctz** [comment]
> if webpki-roots would be willing to implement a new API based on that.

certainly.

in the short term, the other thought i had was finding some root popularity data from somewhere (ssl observatory?) and ordering their appearance in webpki-roots by that.

**Comment by briansmith** [comment]
> in the short term, the other thought i had was finding some root popularity data from somewhere (ssl observatory?) and ordering their appearance in webpki-roots by that.

Yes, that would also be a reasonable way to go. I believe recently there was a study that showed that the top 17 or 19 or so roots were sufficient for ~99% of end users, but I can't find it now.

**Comment by avadacatavra** [comment]
@briansmith IIRC, the root store I'm using has ~153 roots

**Comment by briansmith** [comment]
According to the newer post, https://simrangujral.github.io/Rustls&Servo, “Rustls was always faster for certificate verification.” So, maybe this isn't a priority now?

OTOH, I'm (pleasantly) surprised by the results in that blog post. I definitely expected that we'd need to do more performance work. I think we should make sure we can reproduce those results to ensure they are accurate.

**Comment by avadacatavra** [comment]
@briansmith one of the goals of Simran's work this summer was to see what performance work was necessary in order to get webpki into Servo--we were also pleasantly surprised.  There's still a bit of work to do before we can land her PR, part of that being more performance testing, but I agree that this is no longer a priority.

**Comment by briansmith** [comment]
Thanks for the update @avadacatavra!
